File: src/main/java/com/refinedmods/refinedstorage/inventory/item/validator/PatternItemValidator.java
package com.refinedmods.refinedstorage.inventory.item.validator;

import com.refinedmods.refinedstorage.RSItems;
import com.refinedmods.refinedstorage.api.autocrafting.ICraftingPatternProvider;
import com.refinedmods.refinedstorage.item.PatternItem;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;

import java.util.function.Predicate;

public class PatternItemValidator implements Predicate<ItemStack> {
    private final Level level;

    public PatternItemValidator(Level level) {
        this.level = level;
    }

    @Override
    public boolean test(ItemStack stack) {
        if (stack.getItem() == RSItems.PATTERN.get()) {
            return PatternItem.fromCache(level, stack).isValid();
        }
        return stack.getItem() instanceof ICraftingPatternProvider && ((ICraftingPatternProvider) stack.getItem()).create(level, stack, null).isValid();
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/inventory/item/validator/PatternItemValidator.kt
package com.refinedmods.refinedstorage.inventory.item.validator

import com.refinedmods.refinedstorage.api.autocrafting.ICraftingPatternProvider
import net.minecraft.item.ItemStack
import net.minecraft.world.World
import java.util.function.Predicate

class PatternItemValidator(private val world: World) : Predicate<ItemStack> {
    override fun test(stack: ItemStack): Boolean {
        return stack.item is ICraftingPatternProvider && (stack.item as ICraftingPatternProvider).create(world, stack, null)!!.isValid()
    }
}